examples/camera/emotion_test.py

```diff
@@ -0,0 +1,23 @@
+from pitop.processing.algorithms.faces import FaceDetector, EmotionDetector
+from pitop.processing.core.vision_functions import import_opencv
+
+
+cv2 = import_opencv()
+
+face_detector = FaceDetector()
```

Comment:
Spitballing here. I think we'd aspire to:
face_detector = Detector(face_model)
dog_detector = Detector(dog_model)
it's more flexible, and allows us to explore models more easily with students.
I haven't reviewed FaceDetector yet, so may change / delete this comment later :)
Anyone else have thoughts?

Reply:
something like `face_detector = ObjectDetector(detectors.FACE)` could work, but all it would be doing is giving them the `FaceDetector` class (or `BallDetector`). The issue is that each one will have a very different implementation, it's not just a case of passing a `model` to it

Reply:
let's have a chat - I need to understand that better. Do we have BallDetector already?

Reply:
I think the important thing is not to share implementations but provide a consistent interface to our image processing & feature detector functionality, so that they could be composed in flexible ways.

For example feature detectors should take an image and return a standardised result including a processed image ('robot view') and an array of features, with their positions in the image and confidence values. Those features can then include the non-standard information such as an instance of a 'Face' class with features etc. This should be applicable to our ball, line and face detectors and will help people to use them outside the provided example code because they follow a simple pattern.

Similarly I don't think it's correct for our ball and face detectors to be classes whilst line detector is a function. None of them require tracking previous frame or other state that I can see, so they don't need to be classes, though perhaps it's worth using a class for the one-time configuration and fps tracking features (which certainly could be abstracted into a Detector super class). Motion detection is an example where tracking state from previous frames is required.

This organisation helps to gain insight for ourselves and users for example that emotion detection doesn't fit this pattern the same way as these other detectors. It doesn't use the image as part of it's processing, only for annotation, so perhaps we consider it to be a 'classifier' instead, possibly embedded in the face detector as it always requires that.

Reply:
Both face detector and emotion detector require state, line detector should be changed to a class too as then we can use a Kalman filter to track the line’s position even when it goes slightly off frame - this will also help to minimise false detection too and generally make it more robust.

I agree with the other comments though, standardised return data structure would be nice - I’ve seen this done if other libs that we can take inspiration from. The reason I didn’t embed emotion detector in face detector is because I’d like to have other things like a “drowsiness detector” that also takes a face as input (also needs to be a class since they’ll be filtering there too). I guess there’s multiple ways to do this though so would love to discuss it more